rfcs/0097-no-read-store-dir.md

```diff
@@ -0,0 +1,119 @@
+
+For non-sandboxed programs, much will not change if they can still read /run/current-system, /nix/var,
+etc., thus this change is only important for sandboxes where you can remove such information leaks.
+It isn't a big improvement by itself, but it is a small incremental hardening with few drawbacks that allows for better security in combination with sandboxes.
+
+NB: While this change is simple, it is not possible for end-users to do without changing Nix, since at the moment any
+`nix` command will reset the permissions back to 1775.
+
+# Detailed design
+[design]: #detailed-design
+
+Set the 1775
+in [nixpkgs/nixos/modules/system/boot/stage-2-init.sh](https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/system/boot/stage-2-init.sh#L62),
+in [nix/scripts/install-multi-user.sh](https://github.com/NixOS/nix/blob/cf1d4299a8fa8906f62271dcd878018cef84cc30/scripts/install-multi-user.sh#L577),
+in [nix/src/libstore/globals.hh](https://github.com/NixOS/nix/blob/ba8b39c13003c8ddafb6bec308997e09b9851c46/src/libstore/globals.hh#L278),
+in [nix/src/libstore/build/local-derivation-goal.cc](https://github.com/NixOS/nix/blob/6182ae689826554d915b4ed72e07f7978dc1d13c/src/libstore/build/local-derivation-goal.cc#L641), and
+in [nix/src/libstore/local-store.cc](https://github.com/NixOS/nix/blob/0a535dd5ac93576f7152d786464e330ae3d46b50/src/libstore/local-store.cc#L181)
+to 1771.
+
+/nix/store/trash and /nix/store/.links will also have to have their read bit removed, resulting in 0751.
+
+Currently, /proc/cmdline provides the path to the current system configuration, which is counter-productive in this case.
+This would be fixed by simply setting the permissions to 0440, since the permissions are universal to all pid namespaces.
+Fixing the permissions for other files in /proc would likely also be a good idea, regardless of this RFC and regardless
+of sandboxing, e.g. it doesn't make sense to expose /proc/config.gz to the entire world even if it doesn't expose any secrets.
+
+Nothing else has to be done likely, since setting the store permissions to `1771` manually doesn't break
+anything other than what is mentioned in this document (though it is undone if you run a `nix` command).
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Losing the read (r) bit means that you can't list the files inside the store.
+The execute (x) bit allows us to `cd` to it and also access paths inside the store.
+
+E.g. `ls "$(readlink /nix/var/nix/profiles/system)"` will still work, since this is a directory
+inside the store, and not the store itself, but an unprivileged user can't `ls /nix/store` to find the system configuration.
+
+Note: A sandboxed program could still have some idea of how the host machine is used by checking each store path that Hydra has ever built,
+then it could estimate whether the host machine has e.g. Tor on it.
+
+# Drawbacks
```

Comment:
1. Some users will find it unacceptable that sandboxed programs will be able to probe for specific derivations. For example by probing for common source archives in the store the sandboxed application can likely identify a lot about the host system. This means that this level of access control is not sufficient for some use cases.
2. Leaks are easy to add and may largely remove the benefits of this approach. This makes it a somewhat weak solution as it will be easy to accidentally break the protection and there is no good way to audit against these leaks.

Reply:
I'm not sure I agree with 2. but 1 is true yeah, for some use cases it isn't enough.
Let's say we're using bwrap to sandbox a program, for example some HTTP server, like so:
```
  env - bwrap
    --unshare-all
    --ro-bind /nix/store /nix/store
    --share-net
    --proc /proc
    --dev /dev
    --dir /tmp
    --setenv TMPDIR /tmp
    --ro-bind /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
    --ro-bind /etc/ssl/certs/ca-bundle.crt /etc/ssl/certs/ca-bundle.crt
    --ro-bind /etc/resolv.conf /etc/resolv.conf
    --hostname localhost
    --die-with-parent
    --new-session
    --bind . /data
    --chdir /data
    -- simple-http-server
```
Let's define a leak as the path to a built system configuration, or some similarly "large" derivation such that the closure is essentially everything of importance in the store.
Besides some files in /proc that can be fixed by fixing the permissions, how do you think you could unintentionally expose a leak?

Reply:
Files in `/etc` can link to fairly large closures, kernel APIs such as `/proc` can also have lots of info, binaries made available to the service may contain this information. The focus of my point is that we have no good way of detecting leaks, this makes me weary of relying on this for fairly untrusted applications.

For example `/proc/*/task/*/mountinfo` is somewhat dangerous and hard to just fix the permissions. Even things like `/proc/sys/kernel/modprobe` could be updated by a user to do something like logging and surprisingly reference a secret.

I'm not saying that this is a blocker. But I think they are important concerns to list so that their cost can be weighed against the benefit of the change.

Reply:
We're considering this feature in the context of sandboxes, so let's continue from the example I made above (which albeit not completely secure is still good enough for this example).

We don't have access to the entirety of `/etc` in our sandbox, it doesn't make sense in any case. We do not want sandboxed programs to have access to anything in `/etc` that is not relevant to the sandbox. As such, we simply do not mount it.
As for `/proc/*/task/*/mountinfo`, that will inside the sandbox only show what is mounted inside the sandbox.

`/proc/sys` does indeed contain a couple of paths (notably to systemd and modprobe),  but for the sake of argument,
I will assume that there is some global environment that can not be hidden from sandboxes.
The problem is then that the user might not be aware that adding a path to this hypothetical environment will make it accessible to sandboxed programs, which are perhaps handled transparently by NixOS. Then the user could inadvertently add the path to something important, partially ruining the point of sandboxing. However, you could also argue that this behavior would actually be preferred, since what has access to the path should be able to read the path intuitively.

I can not actually think of any such global state that the user would want to add a path to that isn't `/proc/cmdline`, `/proc/config.gz`, etc., but of course that doesn't exclude it from existing.

I will add something about this tomorrow after I wake up (night here in Denmark right now).

A bit off-topic:
Regardless of this RFC, I think it's unfortunate that leaking information into sandboxes is so trivial in Linux.
`/proc/sys/kernel/modprobe` is of no relevance to most programs, yet it is present in `/proc`, which is universally available AFAIK (although namespaced). systemd does have a way of of restricting access to the non-PID related parts of `/proc` through `ProcSubset`, though it says this about it in the man page:
> Since these APIs are used frequently this option is useful only in a few, specific cases, and is not suitable for most non-trivial programs

Reply:
>  As for /proc/*/task/*/mountinfo, that will inside the sandbox only show what is mounted inside the sandbox.

Yes, but if you map any file from a derivation it will reveal the full path, even if it behind a symlink IIRC. My point isn't that it is impossible to use correctly, but that it is easy to accidentally reveal something that you didn't intend to.

> I will add something about this tomorrow after I wake up (night here in Denmark right now).

Thanks. Again, I don't think this is a blocker as this change does technically improve security. I just want to make sure that we are also documenting the limitations and downside of this approach so that
1. People are extra careful if they do choose to rely upon this change.
2. People don't blindly assume that this is a sufficient solution to remove whatever previous mechanism they have, but instead consider the tradeoff carefully.

Reply:
I agree, I will add something about it.

Reply:
I consider this resolved.